storage/isolation_manager.rs

```diff
@@ -241,297 +358,317 @@ struct Timeline {
-        //             --> There is a complexity in updating the watermark - we need to scan commit statuses from the current watermark upwards, then update it. What if two commits do this concurrently and race? Is it possible to not make progress?
-        //
-        //     We should just re-think what the watermark guarantees actually are...?
-        //     1) Because we commit optimistically/at the end of a snapshot only, commit slots are only filled once a transaction goes to commit. This means we can at worst be as behind as the current set of 'processing' commits, which is not many.
-        //     2) We do allow pending commits to proceed out of order sometimes - if they have non-intersecting commit records. This means we can't guarantee the watermark is raised after a commit.
-        //        --> first hint: maybe we can just bump the watermark when a commit happens that did not skip any commit records/intersected with all concurrent commits? Seems like most cases will not suit this...
-        //     3) Side note: If we want to preserve causality between transactions by the same user, a new transaction by the same user must wait for the watermark to rise past the recently committed snapshot.
-        //         --> simplest way to do this without retaining causality information on the client is: when a client asks for a snapshot, immediately get _latest_ (possibly pending) commit sequence number. Then wait until watermark rises to meet that and use it.
-        //
-        let (next_sequence_number, windows) = &*self.next_window_and_windows.read().unwrap_or_log();
-        windows
-            .iter()
-            .find_map(|w| {
-                if w.is_finished() {
-                    None
-                } else {
-                    debug_assert!(w.watermark().is_some());
-                    w.watermark()
-                }
-            })
-            .unwrap_or_else(|| SequenceNumber::new(next_sequence_number.number() - U80::new(1)))
+    fn may_free_windows(&self) {
+        let watermark = self.watermark();
+        let can_free_some: bool = {
+            let (next_relative_index, windows) = &*self.next_window_and_windows.read().unwrap_or_log();
+            let start_of_second_window = *next_relative_index - ((windows.len() - 1) * TIMELINE_WINDOW_SIZE) as i64;
+            match windows.front() {
+                None => false,
+                Some(front) => front.get_readers() == 0 && watermark >= start_of_second_window,
+            }
+        };
+        if can_free_some {
+            let (next_relative_index, windows) = &mut *self.next_window_and_windows.write().unwrap_or_log();
+            let mut start_of_next_window = *next_relative_index - ((windows.len() - 1) * TIMELINE_WINDOW_SIZE) as i64;
+            while watermark >= start_of_next_window && windows.front().is_some() && windows.front().unwrap().get_readers() == 0
```

Comment:
this is a specific "clearable" condition right? Maybe this is worth its own function with a nice name, if possible/not introducing complexity